### PR TITLE
DataGrid: load more and virtual scrolling

### PR DESCRIPTION
**Pull request by wanghoppe**

### CODE REVIEW

**Comment by wanghoppe** [comment]
AB#25124895

File: packages/playground/src/demo/display/task-grid/task-data-grid.tsx [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+import { Stack } from "@fluentui/react/lib/Stack";
+import { Checkbox } from "@fluentui/react/lib/Checkbox";
+import { useLoadMore } from "@azure/bonito-ui/lib/hooks";
+import { MessageBar, MessageBarType } from "@fluentui/react/lib/MessageBar";
+
+export const DataGridLoadMoreDemo = () => {
+    const [filter, setFilter] = React.useState<string>("");
+    const [limit, setLimit] = React.useState<number>(10);
+    const [noData, setNoData] = React.useState<boolean>(false);
+    const [isCompat, setIsCompat] = React.useState<boolean>(false);
```
Comment (dpwatrous) [inline packages/playground/src/demo/display/task-grid/task-data-grid.tsx:19]: Typo: Compat -> Compact

File: packages/playground/src/demo/display/task-grid/task-data-grid.tsx [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                        label="No data"
+                        checked={noData}
+                        onChange={() => {
+                            setNoData(!noData);
+                        }}
+                    />
+                </Stack.Item>
+                <Stack.Item grow={0}>
+                    <Checkbox
+                        label="Compat Mode"
```
Comment (dpwatrous) [inline packages/playground/src/demo/display/task-grid/task-data-grid.tsx:81]: Typo: Compat -> Compact

**Review by dpwatrous — CHANGES_REQUESTED** [review]
This is looking really good - I love the tests and demos you added. Let's follow up tomorrow and talk more about the overall design.

Random thought: this seems like a really good use case for some automated performance tests. I'd love to be able to track the performance of the data grid in various scenarios over time. We already have a few tests in Playwright which run a real Electron process. Seems like that might be a good place to put perf tests.

**Comment by dpwatrous** [comment]
Looks like there's a styling issue with errors on the demo page (this definitely wouldn't pass color contrast requirements):

 
Screenshot 2023-09-12 at 11 08 59 AM

File: packages/bonito-ui/src/components/data-grid/data-grid.tsx [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+    selectionMode?: "single" | "multiple" | "none";
+
+    /**
+     * Whether there are more items to load.
+     * If true, the grid will display 10 lines of shimmering when items is empty,
+     * indicating that the datagrid is initial loading, or 3 lines of shimmering
+     * when items is not empty, indicating that the datagrid is loading more data.
+     * If false/undefined, the grid will display "No result" when items is empty.
+     */
+    hasMore?: boolean;
```
[the remark this answers is not in the document]
Comment (wanghoppe) [reply packages/bonito-ui/src/components/data-grid/data-grid.tsx:54]: Agreed, one questions is that `hasMore` actually used as a loading indicator for initial load as well. `hasMore = true` while `items` are empty would indicate that it's initial loading, later when the `items` are populated and set `hasMore = false` meaning the initial load is finished. I am thinking whether we should have separate filed `isLoading` to do this.

[the remark this answers is not in the document]
Comment (dpwatrous) [reply packages/bonito-ui/src/components/data-grid/data-grid.tsx:54]: After some discussion, I'm fine if we leave this as-is for now.

**Comment by wanghoppe** [comment]
> Looks like there's a styling issue with errors on the demo page (this definitely wouldn't pass color contrast requirements):
> 
> Screenshot 2023-09-12 at 11 08 59 AM

I use the latest @fluentui/azure-themes instead, still need to make some adjustment to make the error message bar looked OK.  Also adjust some style to make the shimmering more promient on dark and light theme.
image

File: packages/bonito-ui/package.json [hunk 4]
```diff
@@ -34,7 +34,7 @@
     "test:a11y": "cross-env BE_ENABLE_A11Y_TESTING=true npm run test",
     "test:a11y:watch": "cross-env BE_ENABLE_A11Y_TESTING=true npm run test:watch",
     "lint": "eslint .",
-    "watch": "run-p \"compile:* -- --watch --preserveWatchOutput\"",
+    "watch": "npm run compile:esm -- --watch --preserveWatchOutput",
```
Comment (dpwatrous) [inline packages/bonito-ui/package.json:37]: This looks like it got reverted somehow

File: web/dev-server/resources/i18n/resources.en.json [hunk 5]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (dpwatrous) [inline web/dev-server/resources/i18n/resources.en.json:1]: This file shouldn't be checked in (it's in .gitignore)

**Review by dpwatrous — CHANGES_REQUESTED** [review]

**Review by dpwatrous — APPROVED** [review]

**Pull request merged**